### ISSUE TITLE
training code of SCN-SR

### ISSUE THREAD

**Issue by llyydd007** [huangzehao/SCN_Matlab#2]
Dear zehao
Firstly, thanks very much for the great test code of the paper. As you implemented the training code by caffe, would you please share it on github? If it contains some bugs, would you please send me a piece  as I'm very interested about the paper. It will give me great help for understanding the paper and doing some future work. Thank you very much!@huangzehao

**Comment by huangzehao** [comment]
Hi, llyydd007! Thank you for your interest about my implementation. Actually, I have been finished the training code for a long time. But the code is so messy and I have no time to clear up. I will share it on github soon! 

And I think "Demo_SR_Conv.m" is helpful to understand the network structure in the original paper, you can check it firstly. After understanding the network structure and convolution, it's easy for you to implement the training code in caffe, torch or other deep learning tools.

You can also check my implementation of  "Accurate Image Super-Resolution Using Very Deep Convolutional Networks" (CVPR 2016 Oral Paper)  in caffe-vdsr if you are interest in state-of-the-art super-resolution work.

**Comment by llyydd007** [comment]
Thank you very much! I have read the "Demo_SR_Conv.m" and analyzed it. It is simple and clear which easily be converted into caffe architecture. I just wonder if any recurrent part existed in the training process or it is just a feed-forward network?
 I am very interested to the implementation of  rnn in caffe for SR task.  Although there is lstm example for captioning task, I am confused.  I have read the paper and your implementation of "Accurate Image Super-Resolution Using Very Deep Convolutional Networks". It is great. But I'm much more interested for the paper "Deeply-Recursive Convolutional Network for Image Super-Resolution". Anyway, it is great to have so many kindly shared code. And thank you again for your kind response and guidance for me.

**Comment by llyydd007** [comment]
Thank you very much! I have read the "Demo_SR_Conv.m" and analyzed it. It is simple and clear which easily be converted into caffe architecture. I just wonder if any recurrent part existed in the training process or it is just a feed-forward network?
 I am very interested to the implementation of  rnn in caffe for SR task.  Although there is lstm example for captioning task, I am confused.  I have read the paper and your implementation of "Accurate Image Super-Resolution Using Very Deep Convolutional Networks". It is great. But I'm much more interested for the paper "Deeply-Recursive Convolutional Network for Image Super-Resolution". Anyway, it is great to have so many kindly shared code. And thank you again for your kind response and guidance for me.

**Comment by llyydd007** [comment]
Thank you very much! I have read the "Demo_SR_Conv.m" and analyzed it. It is simple and clear which easily be converted into caffe architecture. I just wonder if any recurrent part existed in the training process or it is just a feed-forward network?
 I am very interested to the implementation of  rnn in caffe for SR task.  Although there is lstm example for captioning task, I am confused.  I have read the paper and your implementation of "Accurate Image Super-Resolution Using Very Deep Convolutional Networks". It is great. But I'm much more interested for the paper "Deeply-Recursive Convolutional Network for Image Super-Resolution". Anyway, it is great to have so many kindly shared code. And thank you again for your kind response and guidance for me.

**Comment by llyydd007** [comment]
Thank you very much! I have read the "Demo_SR_Conv.m" and analyzed it. It is simple and clear which easily be converted into caffe architecture. I just wonder if any recurrent part existed in the training process or it is just a feed-forward network?
 I am very interested to the implementation of  rnn in caffe for SR task.  Although there is lstm example for captioning task, I am confused.  I have read the paper and your implementation of "Accurate Image Super-Resolution Using Very Deep Convolutional Networks". It is great. But I'm much more interested for the paper "Deeply-Recursive Convolutional Network for Image Super-Resolution". Anyway, it is great to have so many kindly shared code. And thank you again for your kind response and guidance for me.

**Comment by llyydd007** [comment]
Thank you very much! I have read the "Demo_SR_Conv.m" and analyzed it. It is simple and clear which easily be converted into caffe architecture. I just wonder if any recurrent part existed in the training process or it is just a feed-forward network?
 I am very interested to the implementation of  rnn in caffe for SR task.  Although there is lstm example for captioning task, I am confused.  I have read the paper and your implementation of "Accurate Image Super-Resolution Using Very Deep Convolutional Networks". It is great. But I'm much more interested for the paper "Deeply-Recursive Convolutional Network for Image Super-Resolution". Anyway, it is great to have so many kindly shared code. And thank you again for your kind response and guidance for me.

**Comment by llyydd007** [comment]
Thank you very much! I have read the "Demo_SR_Conv.m" and analyzed it. It is simple and clear which easily be converted into caffe architecture. I just wonder if any recurrent part existed in the training process or it is just a feed-forward network?
 I am very interested to the implementation of  rnn in caffe for SR task.  Although there is lstm example for captioning task, I am confused.  I have read the paper and your implementation of "Accurate Image Super-Resolution Using Very Deep Convolutional Networks". It is great. But I'm much more interested for the paper "Deeply-Recursive Convolutional Network for Image Super-Resolution". Anyway, it is great to have so many kindly shared code. And thank you again for your kind response and guidance for me.

**Comment by llyydd007** [comment]
Thank you very much! I have read the "Demo_SR_Conv.m" and analyzed it. It is simple and clear which easily be converted into caffe architecture. I just wonder if any recurrent part existed in the training process or it is just a feed-forward network?
 I am very interested to the implementation of  rnn in caffe for SR task.  Although there is lstm example for captioning task, I am confused.  I have read the paper and your implementation of "Accurate Image Super-Resolution Using Very Deep Convolutional Networks". It is great. But I'm much more interested for the paper "Deeply-Recursive Convolutional Network for Image Super-Resolution". Anyway, it is great to have so many kindly shared code. And thank you again for your kind response and guidance for me.

**Comment by llyydd007** [comment]
Thank you very much! I have read the "Demo_SR_Conv.m" and analyzed it. It is simple and clear which easily be converted into caffe architecture. I just wonder if any recurrent part existed in the training process or it is just a feed-forward network?
 I am very interested to the implementation of  rnn in caffe for SR task.  Although there is lstm example for captioning task, I am confused.  I have read the paper and your implementation of "Accurate Image Super-Resolution Using Very Deep Convolutional Networks". It is great. But I'm much more interested for the paper "Deeply-Recursive Convolutional Network for Image Super-Resolution". Anyway, it is great to have so many kindly shared code. And thank you again for your kind response and guidance for me.

**Comment by llyydd007** [comment]
Thank you very much! I have read the "Demo_SR_Conv.m" and analyzed it. It is simple and clear which easily be converted into caffe architecture. I just wonder if any recurrent part existed in the training process or it is just a feed-forward network?
 I am very interested to the implementation of  rnn in caffe for SR task.  Although there is lstm example for captioning task, I am confused.  I have read the paper and your implementation of "Accurate Image Super-Resolution Using Very Deep Convolutional Networks". It is great. But I'm much more interested for the paper "Deeply-Recursive Convolutional Network for Image Super-Resolution". Anyway, it is great to have so many kindly shared code. And thank you again for your kind response and guidance for me.

**Comment by llyydd007** [comment]
Thank you very much! I have read the "Demo_SR_Conv.m" and analyzed it. It is simple and clear which easily be converted into caffe architecture. I just wonder if any recurrent part existed in the training process or it is just a feed-forward network?
 I am very interested to the implementation of  rnn in caffe for SR task.  Although there is lstm example for captioning task, I am confused.  I have read the paper and your implementation of "Accurate Image Super-Resolution Using Very Deep Convolutional Networks". It is great. But I'm much more interested for the paper "Deeply-Recursive Convolutional Network for Image Super-Resolution". Anyway, it is great to have so many kindly shared code. And thank you again for your kind response and guidance for me.

**Comment by llyydd007** [comment]
Thank you very much! I have read the "Demo_SR_Conv.m" and analyzed it. It is simple and clear which easily be converted into caffe architecture. I just wonder if any recurrent part existed in the training process or it is just a feed-forward network?
 I am very interested to the implementation of  rnn in caffe for SR task.  Although there is lstm example for captioning task, I am confused.  I have read the paper and your implementation of "Accurate Image Super-Resolution Using Very Deep Convolutional Networks". It is great. But I'm much more interested for the paper "Deeply-Recursive Convolutional Network for Image Super-Resolution". Anyway, it is great to have so many kindly shared code. And thank you again for your kind response and guidance for me.

**Comment by llyydd007** [comment]
Thank you very much! I have read the "Demo_SR_Conv.m" and analyzed it. It is simple and clear which easily be converted into caffe architecture. I just wonder if any recurrent part existed in the training process or it is just a feed-forward network?
 I am very interested to the implementation of  rnn in caffe for SR task.  Although there is lstm example for captioning task, I am confused.  I have read the paper and your implementation of "Accurate Image Super-Resolution Using Very Deep Convolutional Networks". It is great. But I'm much more interested for the paper "Deeply-Recursive Convolutional Network for Image Super-Resolution". Anyway, it is great to have so many kindly shared code. And thank you again for your kind response and guidance for me.

**Comment by llyydd007** [comment]
Thank you very much! I have read the "Demo_SR_Conv.m" and analyzed it. It is simple and clear which easily be converted into caffe architecture. I just wonder if any recurrent part existed in the training process or it is just a feed-forward network?
 I am very interested to the implementation of  rnn in caffe for SR task.  Although there is lstm example for captioning task, I am confused.  I have read the paper and your implementation of "Accurate Image Super-Resolution Using Very Deep Convolutional Networks". It is great. But I'm much more interested for the paper "Deeply-Recursive Convolutional Network for Image Super-Resolution". Anyway, it is great to have so many kindly shared code. And thank you again for your kind response and guidance for me.

**Comment by llyydd007** [comment]
Thank you very much! I have read the "Demo_SR_Conv.m" and analyzed it. It is simple and clear which easily be converted into caffe architecture. I just wonder if any recurrent part existed in the training process or it is just a feed-forward network?
 I am very interested to the implementation of  rnn in caffe for SR task.  Although there is lstm example for captioning task, I am confused.  I have read the paper and your implementation of "Accurate Image Super-Resolution Using Very Deep Convolutional Networks". It is great. But I'm much more interested for the paper "Deeply-Recursive Convolutional Network for Image Super-Resolution". Anyway, it is great to have so many kindly shared code. And thank you again for your kind response and guidance for me.

**Comment by huangzehao** [comment]
Hi, llyydd007. Maybe you will be disappointed. Actually there is not recurrent part in my implementation because in the original paper the k is 1 and it's just a feed-forward network. The author said that they did not find big improvement when they increase k. 
In my opinion, increase k maybe help because of the huge development of very deep neural network training and rnn in 2015. 
And in  "Deeply-Recursive Convolutional Network for Image Super-Resolution", it's also a feed-forward network because the recursive part can be unfolded as a feed-forward network with shared weights. I think the implementation of DRCN is not very difficult. You can just consider it as a feed-forward network. The difficult of implementing it in caffe is the loss function and training.
The training time of DRCN is too long (days in the original paper) and I have no time to try to implement it. If you have interest about recurrent in SR, you can check another paper submitted to ECCV 16, http://arxiv.org/pdf/1604.08671.pdf.  But DRCN and VDSR are still state-of-the-art.

**Comment by llyydd007** [comment]
Thank you very much! It is kind of you to give these advice. Thank you!@huangzehao
